digitiser_decode.cpp

```diff
@@ -69,6 +71,126 @@ static std::vector<std::int16_t> decode_10bit(const std::uint8_t *data, std::siz
     return out;
 }
 
+template<int left>
+[[gnu::target("avx2")]]
+static inline __m256i extr(__m256i in)
+{
+    static_assert(0 <= left && left <= 22, "left is out of ange");
+    if (left)
+        return _mm256_srai_epi32(_mm256_slli_epi32(in, left), 22);
+    else
+        return _mm256_srai_epi32(in, 22);
+}
+
+template<int left>
+[[gnu::target("avx2")]]
+static inline __m256i extr2(__m256i in0, __m256i in1)
+{
+    static_assert(22 < left && left < 32, "left is out of ange");
+    __m256i part = _mm256_srai_epi32(_mm256_slli_epi32(in0, left), 22);
+    return _mm256_or_si256(part, _mm256_srli_epi32(in1, 54 - left));
+}
+
+// AVX-2 optimised version of the above
+[[gnu::target("avx2")]]
+static std::vector<std::int16_t> decode_10bit(const std::uint8_t *data, std::size_t length)
+{
+    std::size_t out_length = length * 8 / 10;
+    std::vector<std::int16_t> out(out_length);
+
+    std::size_t blocks = out_length / 128;
+    // shuffle control to swap 32-bit values from big to little endian
+    __m256i bswap = _mm256_set_epi32(
+        0x0c0d0e0f, 0x08090a0b, 0x04050607, 0x00010203,
+        0x0c0d0e0f, 0x08090a0b, 0x04050607, 0x00010203
+    );
+    const std::int32_t *x_ptr = (const std::int32_t *) data;
+    __m128i *y_ptr = (__m128i *) out.data();
+    __m256i offsets = _mm256_set_epi32(35, 30, 25, 20, 15, 10, 5, 0);
```

Comment:
It would help to specify the point of these offsets. I finally gathered (pardon the pun) that these are counting 32-bit ints, and 5 * 32 = 160 bits is the size of the most basic unit divisible by all the important numbers (32 and 10).

We go up to 40 ints = 1280 bits = the block size. And then we get 5 different interleaved patterns in which the 10 bits are found within the integers.